proposals/interceptors-issues-2024-01.md

```diff
@@ -0,0 +1,152 @@
+
+## File path portability
+
+One of our design goals with interceptors was for them to conform to the "portability" requirement of source generators. That means we should be able to run the source generators for a project containing interceptors, commit the results to disk, then remove the source generators from the project and be able to build in a variety of environments. Note that this is different from *determinism*--we don't mind if the resulting assembly differs in things like layout, etc. in each build environment, but want it to be the case that if an interceptor declaration functions in one environment, then it also functions in a different environment, assuming all the sources and dependencies are fully available in each environment.
+
+However, we currently require absolute paths for `InterceptsLocationAttribute`, with pathmap substitution performed if present. This doesn't work between local builds on different paths, or across local builds and CI, because only CI will actually pass `/pathmap` during the build. The local build prefers to preserve local paths (e.g. in the PDB) so that the local versions of the sources can be discovered easily during debugging. The different environments will have different answers for what they expect the paths to be, and the InterceptsLocation will have errors in environments which differ from the "generation-time" environment.
+
+Having looked at several strategies to address here, including adjusting when the build system passes `/pathmap` to the compiler, it seems the most straightforward way is to permit relative paths. This is a strategy which is already proven out by `#line` directives.
+
+When a relative path appears in an `[InterceptsLocation]`, we will resolve it relative to the path of the containing file, just as we would for `#line`. We recommend that relative paths appearing in source use `/` directory separators for portability.
+
+```cs
+// /project/src/file1.cs
+class Original
+{
+    static void Interceptable() { }
+
+    void M()
+    {
+        Interceptable();
+    }
+}
+
+// /project/obj/generated/file2.g.cs
+class Interceptors
+{
+    [InterceptsLocation("../../src/file1.cs", ...)] // resolves to "/project/src/file1.cs"
+    public static void Interceptor() { }
+}
+```
+
+Since we want generated files to refer relatively to "user" files, we now need to define the paths of generated files. Also, for source generators to be able to insert the correct relative paths into the generated source, a source generator needs to know what the file path will be before the file is actually added to the compilation. We may want to add new public API like the following for this:
+
+```diff
+ namespace Microsoft.CodeAnalysis;
+
+ public readonly struct SourceProductionContext
+ {
+     public void AddSource(string hintName, string source);
++    public string GetFilePath(string hintName);
++    public string GetInterceptsFilePath(SyntaxNode intercepted, string hintName);
```

Comment:
Perhaps the public API should just give "the stuff to stick in the InterceptsLocationAttribute". It's possible that if we move toward an "opaque identifier" approach, and the analyzer author is using such an API, that the compiler could "transparently" update the behavior of the API, in order to automatically start emitting new "versions" of the opaque representation, which might have benefits such as aiding incrementality/delayed generator runs in the IDE.

We would then have to also discuss the concept of versioning the opaque representation, probably with some kind of simple prefix, and the stipulation that older "versions" would work in newer compilers, but older compilers won't be able to decode newer "versions".